File: web/src/p2p/frames.ts
// 隧道帧编解码：与后端 internal/p2p/frames.go 对齐
// 逻辑流 = [4B BE msgLen][envJSON][bin]；>16KB 拆多 packet，每 packet 带 frag 头。
export const FRAG_LIMIT = 16000

export interface Envelope {
  v?: string
  t: string
  id?: number
  m?: string
  p?: string
  q?: Record<string, string>
  h?: Record<string, string>
  b?: unknown // body（j 的请求体／R 的 token，直接给原始值，由本层序列化）
  s?: number // size
  c?: number // ack 码 / 上传分片 idx
  msg?: string
  l?: boolean // 下载最后帧
  dl?: number // 下载文件 id
  hash?: string
  idx?: number
  tot?: number
  ts?: number
  ok?: boolean
  status?: number // http 状态
  f?: number // 分片序号
  fn?: number // 分片总数
  off?: number // 区间下载起始偏移
  len?: number // 区间下载长度
}

function buildPacket(frag: Uint8Array, chunk: Uint8Array): Uint8Array {
  const buf = new Uint8Array(8 + frag.length + chunk.length)
  const dv = new DataView(buf.buffer)
  dv.setUint32(0, frag.length)
  buf.set(frag, 4)
  dv.setUint32(4 + frag.length, chunk.length)
  buf.set(chunk, 8 + frag.length)
  return buf
}

/** 编码一条逻辑消息为若干 packet */
export function encodeFrame(env: Envelope, bin?: Uint8Array): Uint8Array[] {
  const envB = new TextEncoder().encode(JSON.stringify(env))
  const stream = new Uint8Array(4 + envB.length + (bin ? bin.length : 0))
  const dv = new DataView(stream.buffer)
  dv.setUint32(0, envB.length)
  stream.set(envB, 4)
  if (bin) stream.set(bin, 4 + envB.length)

  if (stream.length <= FRAG_LIMIT) {
    return [buildPacket(new TextEncoder().encode(JSON.stringify({ t: env.t, id: env.id })), stream)]
  }
  const n = Math.ceil(stream.length / FRAG_LIMIT)
  const out: Uint8Array[] = []
  for (let i = 0; i < n; i++) {
    const start = i * FRAG_LIMIT
    const end = Math.min(start + FRAG_LIMIT, stream.length)
    const frag = new TextEncoder().encode(JSON.stringify({ t: env.t, id: env.id, f: i, fn: n }))
    out.push(buildPacket(frag, stream.subarray(start, end)))
  }
  return out
}

/** 聚合器：把 DataChannel 收到的 packet 重组成完整逻辑消息 */
export class FrameDecoder {
  private partial = new Map<string, { chunks: Uint8Array[]; count: number; need: number }>()

  feed(pkt: Uint8Array): { env: Envelope; bin: Uint8Array } | null {
    if (pkt.length < 8) return null
    const dv = new DataView(pkt.buffer, pkt.byteOffset, pkt.byteLength)
    const lj = dv.getUint32(0)
    if (pkt.length < 4 + lj + 4) return null
    let frag: Envelope
    try {
      frag = JSON.parse(new TextDecoder().decode(pkt.subarray(4, 4 + lj)))
    } catch {
      return null
    }
    const lc = dv.getUint32(4 + lj)
    if (pkt.length < 8 + lj + lc) return null
    const chunk = pkt.subarray(8 + lj, 8 + lj + lc)

    if (!frag.fn) {
      return this.parseStream(chunk)
    }
    const key = `${frag.t}|${frag.id}`
    let e = this.partial.get(key)
    if (!e) {
      e = { chunks: [], count: 0, need: frag.fn }
      this.partial.set(key, e)
    }
    e.chunks.push(chunk)
    e.count++
    if (e.count >= e.need) {
      this.partial.delete(key)
      const total = e.chunks.reduce((s, c) => s + c.length, 0)
      const stream = new Uint8Array(total)
      let off = 0
      for (const c of e.chunks) {
        stream.set(c, off)
        off += c.length
      }
      return this.parseStream(stream)
    }
    return null
  }

  private parseStream(stream: Uint8Array): { env: Envelope; bin: Uint8Array } | null {
    if (stream.length < 4) return null
    const dv = new DataView(stream.buffer, stream.byteOffset, stream.byteLength)
    const msgLen = dv.getUint32(0)
    if (stream.length < 4 + msgLen) return null
    let env: Envelope
    try {
      env = JSON.parse(new TextDecoder().decode(stream.subarray(4, 4 + msgLen)))
    } catch {
      return null
    }
    return { env, bin: stream.subarray(4 + msgLen) }
  }
}

File: backend/internal/p2p/frames.go
package p2p

import (
	"encoding/binary"
	"encoding/json"
	"strconv"
)

// fragLimit 单条 DataChannel 消息上限。由于浏览器端上限不一（Chromium 256KB、Firefox 16KB、
// Safari 更小），所有跨端传输都必须按此阈值分帧，接收端按 (t, id) 聚合重组。
const fragLimit = 16000

// SDP 值：信令中使用的会话描述
type SDP struct {
	Type string `json:"type"`
	SDP  string `json:"sdp"`
}

// Cand ICECandidate 值
type Cand struct {
	Candidate     string `json:"candidate"`
	SDPMid        string `json:"sdpMid"`
	SDPMLineIndex uint16 `json:"sdpMLineIndex"`
}

// Envelope 隧道帧信封（DataChannel 每条逻辑消息的解析头）
// t ∈ R(握手) P/p(心跳) j(JSON请求) J(JSON响应) d(下载开启) D(下载数据) U(上传分片) A(ack/取消) E(错误) C(关闭)
type Envelope struct {
	V      string            `json:"v,omitempty"`
	T      string            `json:"t"`      // 类型
	ID     uint32            `json:"id,omitempty"`  // 请求/流关联 id
	M      string            `json:"m,omitempty"`   // http 方法（j）
	P      string            `json:"p,omitempty"`   // http 路径（j）
	Q      map[string]string `json:"q,omitempty"`   // query（j）
	H      map[string]string `json:"h,omitempty"`   // headers（j）
	B      json.RawMessage   `json:"b,omitempty"`   // body（j/J 的 JSON body；R 持 token）
	S      int64             `json:"s,omitempty"`   // 大小（U 分片 size / 表示期待长度）
	C      int               `json:"c,omitempty"`   // ack/err 码 / 上传分片 idx
	Msg    string            `json:"msg,omitempty"` // 错误信息
	L      bool              `json:"l,omitempty"`   // 下载最后帧
	DL     uint64            `json:"dl,omitempty"`  // 下载文件 id
	Hash   string            `json:"hash,omitempty"` // 上传分片 hash
	IDX    int               `json:"idx,omitempty"` // 上传分片 index
	TOT    int               `json:"tot,omitempty"` // 上传总片数
	TS     int64             `json:"ts,omitempty"`  // ping/pong 时间戳
	OK     bool              `json:"ok,omitempty"`  // 握手成功
	Status int               `json:"status,omitempty"` // http 状态（J）
	F      int               `json:"f,omitempty"`     // 分片序号（重组用）
	FN     int               `json:"fn,omitempty"`    // 分片总数（重组用）
	SDP    *SDP              `json:"sdp,omitempty"`   // 信令 offer/answer
	Cand   *Cand             `json:"cand,omitempty"`  // 信令 ICE
	CID    string            `json:"cid,omitempty"`   // 信令客户端 id
	To     string            `json:"to,omitempty"`    // 信令目标（单播）
	Off    int64             `json:"off,omitempty"`   // 区间下载的起始字节偏移（>=0 时启用）
	Len    int64             `json:"len,omitempty"`   // 区间下载的长度上限（-1 = 到文件末尾，0 = 完整下载不指定）
}

// packet 单条 DataChannel 消息：header(4B lenFrag) + fragJSON + (4B lenChunk) + chunkBytes
func buildPacket(frag []byte, chunk []byte) []byte {
	buf := make([]byte, 0, 8+len(frag)+len(chunk))
	var hb [4]byte
	binary.BigEndian.PutUint32(hb[:], uint32(len(frag)))
	buf = append(buf, hb[:]...)
	buf = append(buf, frag...)
	binary.BigEndian.PutUint32(hb[:], uint32(len(chunk)))
	buf = append(buf, hb[:]...)
	buf = append(buf, chunk...)
	return buf
}

// frameEncode 将一条逻辑消息（env + bin）编码为若干 packet。
// 逻辑流统一为 [4B msgLen][msgJSON][bin]，按 fragLimit 分帧；SCTP 有序，接收端按 FN 计数聚合。
func frameEncode(env Envelope, bin []byte) [][]byte {
	msg, _ := json.Marshal(env)
	var h [4]byte
	binary.BigEndian.PutUint32(h[:], uint32(len(msg)))
	stream := append(h[:], msg...)
	stream = append(stream, bin...)
	if len(stream) <= fragLimit {
		frag := Envelope{T: env.T, ID: env.ID}
		fj, _ := json.Marshal(frag)
		return [][]byte{buildPacket(fj, stream)}
	}
	n := (len(stream) + fragLimit - 1) / fragLimit
	out := make([][]byte, 0, n)
	for i := 0; i < n; i++ {
		lo := i * fragLimit
		hi := lo + fragLimit
		if hi > len(stream) {
			hi = len(stream)
		}
		frag := Envelope{T: env.T, ID: env.ID, F: i, FN: n}
		fj, _ := json.Marshal(frag)
		out = append(out, buildPacket(fj, stream[lo:hi]))
	}
	return out
}

// fragments 按 (t,id) 聚合从 DataChannel 收到的 packet
type fragments struct {
	env    Envelope
	chunks [][]byte
	count  int
	need   int
}

// assembler 聚合器：收到 packet 后返回已完成的一条完整逻辑消息流
type assembler struct {
	partial map[string]*fragments // key: t|id
}

func newAssembler() *assembler { return &assembler{partial: map[string]*fragments{}} }

// feed 处理一个 packet，返回已聚合完成的逻辑消息流；未完成时返回 (nil, false)。
func (a *assembler) feed(pkt []byte) ([]byte, bool) {
	if len(pkt) < 8 {
		return nil, false
	}
	lj := int(binary.BigEndian.Uint32(pkt[0:4]))
	if lj < 0 || len(pkt) < 4+lj+4 {
		return nil, false
	}
	var frag Envelope
	if err := json.Unmarshal(pkt[4:4+lj], &frag); err != nil {
		return nil, false
	}
	lc := int(binary.BigEndian.Uint32(pkt[4+lj : 8+lj]))
	if len(pkt) < 8+lj+lc {
		return nil, false
	}
	chunk := pkt[8+lj : 8+lj+lc]

	// 非分片：chunk 即完整逻辑流（含 4B msgLen 前缀）
	if frag.FN == 0 {
		return chunk, true
	}

	key := frag.T + "|" + strconv.Itoa(int(frag.ID))
	f := a.partial[key]
	if f == nil {
		f = &fragments{env: frag, need: frag.FN, count: 0}
		a.partial[key] = f
	}
	f.chunks = append(f.chunks, chunk)
	f.count++
	if f.count >= f.need {
		stream := f.chunks[0]
		for _, c := range f.chunks[1:] {
			stream = append(stream, c...)
		}
		delete(a.partial, key)
		return stream, true
	}
	return nil, false
}

// parseStream 从完整逻辑流解析出信封与剩余二进制载荷
func parseStream(stream []byte) (Envelope, []byte, bool) {
	if len(stream) < 4 {
		return Envelope{}, nil, false
	}
	msgLen := int(binary.BigEndian.Uint32(stream[0:4]))
	if msgLen < 0 || len(stream) < 4+msgLen {
		return Envelope{}, nil, false
	}
	var env Envelope
	if err := json.Unmarshal(stream[4:4+msgLen], &env); err != nil {
		return Envelope{}, nil, false
	}
	return env, stream[4+msgLen:], true
}